File: app/src/main/java/app/familygem/GalleryFragment.java
package app.familygem;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.ContextMenu;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import com.theartofdev.edmodo.cropper.CropImage;
import org.folg.gedcom.model.Media;
import org.folg.gedcom.model.MediaContainer;
import org.folg.gedcom.model.MediaRef;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Set;
import app.familygem.visitor.MediaListContainer;
import app.familygem.visitor.MediaReferences;
import app.familygem.visitor.FindStack;
import static app.familygem.Global.gc;

/**
 * List of Media
 * */
public class GalleryFragment extends Fragment {

	MediaListContainer mediaVisitor;
	MediaGalleryAdapter adapter;

	@Override
	public View onCreateView( LayoutInflater inflater, ViewGroup container, Bundle bandolo ) {
		setHasOptionsMenu( true );
		View view = inflater.inflate( R.layout.galleria, container, false );
		RecyclerView recyclerView = view.findViewById( R.id.galleria );
		recyclerView.setHasFixedSize( true );
		if( gc != null ) {
			mediaVisitor = new MediaListContainer( gc, !getActivity().getIntent().getBooleanExtra("galleriaScegliMedia",false ) );
			gc.accept(mediaVisitor);
			setToolbarTitle();
			RecyclerView.LayoutManager gestoreLayout = new GridLayoutManager( getContext(), 2 );
			recyclerView.setLayoutManager( gestoreLayout );
			adapter = new MediaGalleryAdapter( mediaVisitor.mediaList, true );
			recyclerView.setAdapter(adapter);
			view.findViewById( R.id.fab ).setOnClickListener( v ->
					F.displayImageCaptureDialog( getContext(), GalleryFragment.this, 4546, null )
			);
		}
		return view;
	}

	/**
	 * Leaving the activity resets the extra if no shared media has been chosen
	 * // Andandosene dall'attività resetta l'extra se non è stato scelto un media condiviso
	 * */
	@Override
	public void onPause() {
		super.onPause();
		getActivity().getIntent().removeExtra("galleriaScegliMedia");
	}

	void setToolbarTitle() {
		((AppCompatActivity)getActivity()).getSupportActionBar().setTitle( mediaVisitor.mediaList.size()
				+ " " + getString(R.string.media).toLowerCase() );
	}

	/**
	 * Update the contents of the gallery
	 * */
	void recreate() {
		mediaVisitor.mediaList.clear();
		gc.accept(mediaVisitor);
		adapter.notifyDataSetChanged();
		setToolbarTitle();
	}

	// todo bypassabile?
	static int popularity(Media med ) {
		MediaReferences riferiMedia = new MediaReferences( gc, med, false );
		return riferiMedia.num;
	}

	static Media newMedia(Object container ){
		Media media = new Media();
		media.setId( U.newID(gc,Media.class) );
		media.setFileTag("FILE"); // Necessary to then export the Gedcom
		gc.addMedia( media );
		if( container != null ) {
			MediaRef mediaRef = new MediaRef();
			mediaRef.setRef( media.getId() );
			((MediaContainer)container).addMediaRef( mediaRef );
		}
		return media;
	}

	/**
	 * Detach a shared media from a container
	 * */
	static void disconnectMedia(String mediaId, MediaContainer container) {
		Iterator<MediaRef> refs = container.getMediaRefs().iterator();
		while( refs.hasNext() ) {
			MediaRef ref = refs.next();
			if( ref.getMedia( Global.gc ) == null // Possible ref to a non-existent media
					|| ref.getRef().equals(mediaId) )
				refs.remove();
		}
		if( container.getMediaRefs().isEmpty() )
			container.setMediaRefs( null );
	}

	/**
	 * // Delete a shared or local media and remove references in containers
	 * // Return an array with modified progenitors
	 *
	 * // Elimina un media condiviso o locale e rimuove i riferimenti nei contenitori
	 * // Restituisce un array con i capostipiti modificati
	 * */
	public static Object[] deleteMedia(Media media, View view) {
		Set<Object> heads;
		if( media.getId() != null ) { // media OBJECT
			gc.getMedia().remove(media);
			// Delete references in all containers
			MediaReferences deleteMedia = new MediaReferences(gc, media, true);
			heads = deleteMedia.founders;
		} else { // media LOCALE
			new FindStack(gc, media); //temporarily find the media stack to locate the container // trova temporaneamente la pila del media per individuare il container
			MediaContainer container = (MediaContainer) Memory.getSecondToLastObject();
			container.getMedia().remove(media);
			if( container.getMedia().isEmpty() )
				container.setMedia(null);
			heads = new HashSet<>(); // set with only one parent Object
			heads.add( Memory.firstObject() );
			Memory.clearStackAndRemove(); // delete the stack you just created
		}
		Memory.setInstanceAndAllSubsequentToNull(media);
		if( view != null )
			view.setVisibility(View.GONE);
		return heads.toArray(new Object[0]);
	}

	/**
	 * The file fished by the file manager becomes shared media
	 * // Il file pescato dal file manager diventa media condiviso
	 * */
	@Override
	public void onActivityResult( int requestCode, int resultCode, Intent data ) {
		if( resultCode == Activity.RESULT_OK ) {
			if( requestCode == 4546 ) { //File taken from the supplier app is saved in Media and possibly cropped // File preso da app fornitrice viene salvato in Media ed eventualmente ritagliato
				Media media = newMedia(null);
				if( F.proposeCropping(getContext(), this, data, media) ) { // if it is an image (therefore it can be cropped)
					U.save(false, media);
							//onRestart () + recreate () must not be triggered because then the arrival fragment is no longer the same // Non deve scattare onRestart() + recreate() perché poi il fragment di arrivo non è più lo stesso
					return;
				}
			} else if( requestCode == CropImage.CROP_IMAGE_ACTIVITY_REQUEST_CODE ) {
				F.endImageCropping(data);
			}
			U.save(true, Global.croppedMedia);
		} else if( requestCode == CropImage.CROP_IMAGE_ACTIVITY_REQUEST_CODE ) // if you click the back arrow in Crop Image
			Global.edited = true;
	}

	// contextual Menu
	private Media media;
	@Override
	public void onCreateContextMenu( ContextMenu menu, View view, ContextMenu.ContextMenuInfo info ) {
		media = (Media) view.getTag( R.id.tag_object );
		menu.add(0, 0, 0, R.string.delete );
	}
	@Override
	public boolean onContextItemSelected(MenuItem item) {
		if( item.getItemId() == 0 ) {
			Object[] modified = deleteMedia(media, null);
			recreate();
			U.save(false, modified);
			return true;
		}
		return false;
	}

	@Override
	public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
		menu.add(0, 0, 0, R.string.media_folders);
	}
	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		if( item.getItemId() == 0 ) {
			startActivity(new Intent(getContext(), MediaFolders.class)
					.putExtra("idAlbero", Global.settings.openTree)
			);
			return true;
		}
		return false;
	}

	@Override
	public void onRequestPermissionsResult(int codice, @NonNull String[] permission, @NonNull int[] grantResults) {
		F.permissionsResult(getContext(), this, codice, permission, grantResults, null);
	}
}

File: app/src/main/java/app/familygem/main/GalleryFragment.kt
package app.familygem.main

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.ContextMenu
import android.view.ContextMenu.ContextMenuInfo
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.webkit.URLUtil
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.ActionBar
import androidx.appcompat.widget.SearchView
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import app.familygem.FileUri
import app.familygem.Global
import app.familygem.MediaFoldersActivity
import app.familygem.ProgressView
import app.familygem.R
import app.familygem.constant.Choice
import app.familygem.constant.Destination
import app.familygem.constant.Extra
import app.familygem.util.ChangeUtil
import app.familygem.util.FileUtil
import app.familygem.util.MediaUtil
import app.familygem.util.TreeUtil
import app.familygem.util.Util
import app.familygem.visitor.MediaLeaders
import app.familygem.visitor.MediaReferences
import com.google.android.material.floatingactionbutton.FloatingActionButton
import kotlinx.coroutines.Dispatchers.Default
import kotlinx.coroutines.Dispatchers.IO
import kotlinx.coroutines.Dispatchers.Main
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.coroutines.yield
import org.folg.gedcom.model.Media
import java.io.File
import java.net.HttpURLConnection
import java.net.URL
import java.util.Collections

/** Fragment with a list of all Media of the tree. */
class GalleryFragment : BaseFragment() {

    private lateinit var mediaVisitor: MediaLeaders
    lateinit var adapter: GalleryAdapter
    private var sharedMediaOnly = false
    private lateinit var progress: ProgressView
    private lateinit var treeDir: File // Tree media folder in app external storage
    private var searchView: SearchView? = null
    private var checkJob: Job? = null
    private var copyJob: Job? = null
    private var shrinkJob: Job? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        val view = inflater.inflate(R.layout.recyclerview, container, false)
        val recyclerView = view.findViewById<RecyclerView>(R.id.recycler_view)
        recyclerView.setHasFixedSize(true)
        view.findViewById<FloatingActionButton>(R.id.fab).setOnClickListener {
            FileUtil.displayFileChooser(requireContext(), sharedMediaLauncher, Destination.SHARED_MEDIA)
        }
        sharedMediaOnly = requireActivity().intent.getBooleanExtra(Choice.MEDIA, false)
        mediaVisitor = MediaLeaders(sharedMediaOnly, true)
        recyclerView.layoutManager = GridLayoutManager(context, 2)
        adapter = GalleryAdapter(mediaVisitor.list)
        recyclerView.adapter = adapter
        setupFastScroller(recyclerView)
        progress = view.findViewById(R.id.recycler_progress)
        treeDir = requireContext().getExternalFilesDir(Global.settings.openTree.toString())!!
        return view
    }

    override fun showContent() {
        progress.visibility = View.VISIBLE
        mediaVisitor.list.clear()
        Global.gc.accept(mediaVisitor)
        adapter.filter.filter(searchView?.query ?: "")
        progress.visibility = View.GONE
    }

    override fun isSearching(): Boolean {
        return searchView?.query?.isNotBlank() == true
    }

    override fun onPause() {
        super.onPause()
        // Resets the extra if no shared media has been chosen
        requireActivity().intent.removeExtra(Choice.MEDIA)
        // Cancels any active job
        if (checkJob?.isActive == true) checkJob?.cancel()
        if (copyJob?.isActive == true) copyJob?.cancel()
        if (shrinkJob?.isActive == true) shrinkJob?.cancel()
        progress.hideBar()
        progress.visibility = View.GONE
    }

    /** The file retrieved from SAF becomes a shared media. */
    private val sharedMediaLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
        if (result.resultCode == Activity.RESULT_OK) {
            val sharedMedia = MediaUtil.newSharedMedia(null)
            if (FileUtil.setFileAndProposeCropping(requireContext(), result.data, sharedMedia)) {
                finalizeAction(arrayOf(sharedMedia))
            }
        }
    }

    private lateinit var media: Media

    override fun onCreateContextMenu(menu: ContextMenu, view: View, info: ContextMenuInfo?) {
        media = view.getTag(R.id.tag_object) as Media
        if (media.id != null) {
            val index = Global.gc.media.indexOf(media)
            if (index > 0) menu.add(3, 0, 0, R.string.move_up)
            if (index < Global.gc.media.size - 1) menu.add(3, 1, 0, R.string.move_down)
            val mediaReferences = MediaReferences(Global.gc, media, false)
            if (mediaReferences.num > 0) menu.add(3, 2, 0, R.string.make_media)
        } else menu.add(3, 3, 0, R.string.make_shared_media)
        menu.add(3, 4, 0, R.string.delete)
    }

    override fun onContextItemSelected(item: MenuItem): Boolean {
        if (item.groupId == 3) {
            return when (item.itemId) {
                0 -> finalizeAction(swapSharedMedia(-1))
                1 -> finalizeAction(swapSharedMedia(1))
                2 -> finalizeAction(MediaUtil.makeSimpleMedia(media))
                3 -> finalizeAction(MediaUtil.makeSharedMedia(media))
                4 -> {
                    Util.confirmDelete(requireContext()) {
                        finalizeAction(MediaUtil.deleteMedia(media))
                    }
                    true
                }
                else -> false
            }
        }
        return false
    }

    private fun swapSharedMedia(direction: Int): Array<Any> {
        val index = Global.gc.media.indexOf(media)
        Collections.swap(Global.gc.media, index, index + direction)
        return emptyArray()
    }

    /** Saves the changes and updates the content. */
    private fun finalizeAction(modified: Array<Any>): Boolean {
        TreeUtil.save(true, *modified)
        showContent()
        (requireActivity() as MainActivity).refreshInterface()
        return true
    }

    override fun updateToolbar(bar: ActionBar, menu: Menu, inflater: MenuInflater) {
        bar.title = "${mediaVisitor.list.size} ${Util.caseString(R.string.media)}"
        if (!sharedMediaOnly) {
            menu.add(0, 0, 0, R.string.media_folders)
            menu.add(0, 1, 0, R.string.copy_app_storage)
            menu.add(0, 2, 0, R.string.shrink_path)
            checkJob = lifecycleScope.launch(IO) {
                // Sometimes long running tasks
                if (!isThereAnyExternalFile(requireContext())) withContext(Main) { menu.removeItem(1) }
                if (!isThereAnyShrinkablePath(requireContext())) withContext(Main) { menu.removeItem(2) }
            }
        }
        // Search for Media
        if (mediaVisitor.list.size > 1) {
            inflater.inflate(R.menu.search, menu)
            searchView = menu.findItem(R.id.search_item).actionView as SearchView
            searchView?.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
                override fun onQueryTextChange(query: String): Boolean {
                    adapter.filter.filter(query)
                    return true
                }

                override fun onQueryTextSubmit(q: String): Boolean {
                    searchView?.clearFocus()
                    return false
                }
            })
        }
    }

    override fun selectItem(id: Int) {
        when (id) {
            0 -> startActivity(Intent(context, MediaFoldersActivity::class.java).putExtra(Extra.TREE_ID, Global.settings.openTree))
            1 -> {
                progress.visibility = View.VISIBLE
                copyJob = lifecycleScope.launch(IO) { copyFilesToTreeStorage() }
            }
            2 -> {
                progress.visibility = View.VISIBLE
                shrinkJob = lifecycleScope.launch(Default) { shrinkMediaPaths() }
            }
        }
    }

    /** @return True in case there is at least one media linked to a file not in the app storage. */
    private fun isThereAnyExternalFile(context: Context): Boolean {
        return mediaVisitor.list.filterNot { it.media.file.isNullOrBlank() }.any {
            if (it.fileUri == null) it.fileUri = FileUri(context, it.media)
            it.fileUri?.file?.startsWith(treeDir) == false || it.fileUri?.uri != null
                    || it.media.file.startsWith("https://") || it.media.file.startsWith("http://")
        }
    }

    /** @return True in case there is at least one media link shrinkable to filename only. */
    private fun isThereAnyShrinkablePath(context: Context): Boolean {
        return mediaVisitor.list.filterNot { it.media.file.isNullOrBlank() }.any {
            if (it.fileUri == null) it.fileUri = FileUri(context, it.media)
            it.fileUri?.treeDirFilename == true
        }
    }

    private var copiedFiles = 0 // Number of files actually copied
    private var toBeSaved = false // Some media link has been modified

    /** Each valid file outside the tree app storage is copied inside. */
    private suspend fun copyFilesToTreeStorage() {
        // Creates a list of media grouped by the file they are linked to
        progress.displayBar("Preparing files", mediaVisitor.list.size.toLong())
        var count: Long = 0
        mediaVisitor.list.forEach {
            yield()
            if (it.fileUri == null) it.fileUri = FileUri(requireContext(), it.media)
            progress.progress = ++count
        }
        progress.hideBar()
        val groupedMedia = mediaVisitor.list.filterNot { it.media.file.isNullOrBlank() || it.fileUri?.file?.startsWith(treeDir) == true }
            .groupBy { it.fileUri?.path ?: it.media.file }
        // Copies each file
        progress.displayBar("Copying files", groupedMedia.size.toLong())
        count = 0
        copiedFiles = 0
        val excluded = arrayOf("text/html", "text/javascript", "application/json", "text/css", "text/xml")
        groupedMedia.forEach { entry ->
            entry.value[0].fileUri?.apply {
                val name = name ?: URLUtil.guessFileName(media.file, null, null)
                val newFile = FileUtil.nextAvailableFileName(treeDir, name)
                if (file != null) {
                    file!!.copyTo(newFile)
                    completeCopy(entry.value, newFile)
                } else if (uri != null) {
                    requireContext().contentResolver.openInputStream(uri!!).use { input ->
                        newFile.outputStream().use { output ->
                            input?.copyTo(output)
                            completeCopy(entry.value, newFile)
                        }
                    }
                } else if (media.file.startsWith("https://") || media.file.startsWith("http://")) {
                    try {
                        val connection = URL(media.file).openConnection() as HttpURLConnection
                        if (connection.responseCode == HttpURLConnection.HTTP_OK && excluded.none { connection.contentType.contains(it) }) {
                            connection.inputStream.use { input ->
                                newFile.outputStream().use { output ->
                                    input?.copyTo(output)
                                    completeCopy(entry.value, newFile)
                                }
                            }
                        } else completeCopy()
                    } catch (ignored: Exception) {
                        ignored.printStackTrace()
                    }
                } else completeCopy()
            }
            progress.progress = ++count
        }
        if (toBeSaved) TreeUtil.save(true)
        withContext(Main) {
            showContent()
            (requireActivity() as MainActivity).refreshInterface()
            progress.hideBar()
            progress.visibility = View.GONE
            val message = if (copiedFiles > 0) "$copiedFiles files copied." else "No file copied."
            Toast.makeText(requireContext(), message, Toast.LENGTH_LONG).show()
        }
    }

    private fun completeCopy(wrapperList: List<MediaLeaders.MediaWrapper>? = null, newFile: File? = null) {
        if (wrapperList != null && newFile != null) {
            val name = newFile.name
            wrapperList.filter { it.media.file != name }.forEach {
                it.media.file = name
                ChangeUtil.updateChangeDate(it.leader)
                toBeSaved = true
            }
            copiedFiles++
        }
    }

    /** Reduces media links to filename only where possible. */
    private suspend fun shrinkMediaPaths() {
        progress.displayBar("Preparing files", mediaVisitor.list.size.toLong())
        var count: Long = 0
        mediaVisitor.list.forEach {
            yield()
            if (it.fileUri == null) it.fileUri = FileUri(requireContext(), it.media)
            progress.progress = ++count
        }
        progress.hideBar()
        val shrinkableMedia = mediaVisitor.list.filter { it.fileUri?.treeDirFilename == true }
        var modified = 0
        shrinkableMedia.forEach { wrapper ->
            wrapper.fileUri?.name?.let {
                wrapper.media.file = it
                ChangeUtil.updateChangeDate(wrapper.leader)
                modified++
            }
        }
        if (modified > 0) TreeUtil.save(true)
        withContext(Main) {
            showContent()
            (requireActivity() as MainActivity).refreshInterface()
            progress.visibility = View.GONE
            val message = if (modified > 0) "$modified paths shrunk." else "No path shrunk."
            Toast.makeText(requireContext(), message, Toast.LENGTH_LONG).show()
        }
    }
}
